### PR TITLE
Sample train/test splitting

### PR DESCRIPTION
**Pull request by jschueller**
hi,

I'm investigating how to allow to split training/testing samples in openturns (and also bootstraping samples)
in the context where we have the input in one sample and the output in another.

For the bootstrapping part I first imagined to create a new "DualBootstrapExperiment" built from 2 samples (X,Y) in which the generate method yields 2 also samples (from the same selection):
```
experiment = ot.DualBootstrapExperiment(X, Y)
x_boot, y_boot = experiment.generate()
```

Now if I need splitting + x/y samples, I'm going to need something with 4 samples, that's sounds bad.

Then I looked at sklearn, they just use indices to generate training/testing samples:
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html
```
kf = KFold(n_splits=2)
for train_index, test_index in kf.split(X):
...     print("TRAIN:", train_index, "TEST:", test_index)
...     X_train, X_test = X[train_index], X[test_index]
...     y_train, y_test = y[train_index], y[test_index]
```

That's more elegant: we can split 1 or n samples, and we already have the operator[] from a list of indices (select in c++).
Also, we can reuse what already exists for the bootstrap in a similar fashion as the BootstrapExperiment provides a method to generate only the selection indices (instead of creating a new class):
```
sel = ot.BootstrapExperiment.GenerateSelection(size, size)
x1, y1 = x[sel], y[sel]

```

So this PR is about adding two new classes LeaveOneOutSplitter and KFoldSplitter that just generate pairs of Indices:
```
size = 10
x = ot.Normal().getSample(size)
#splitter = ot.LeaveOneOutSplitter(size)
splitter = ot.KFoldSplitter(size, 5)
splitter.setRandomize(True)
for indices1, indices2 in splitter:
    x_train, x_test = x[indices1], x[indices2]
    print(indices1, indices2)
[3,7,1,8,2,9,5,0] [4,6]
[4,7,1,8,6,9,5,0] [3,2]
[4,3,1,8,6,2,5,0] [7,9]
[4,3,7,8,6,2,9,0] [1,5]
[4,3,7,1,6,2,9,5] [8,0]
```
I used an iterator method because it's not easy to return a list of pairs of Indices if we want to retrieve everything at once.

I mark it as draft, this is just a proposal, let's hear your comments.

cc @mbaudin47

### CODE REVIEW

**Comment by mbaudin47** [comment]
I must take the `ot.BootstrapExperiment.GenerateSelection` method into account in the example I wrote for #1568

**Comment by mbaudin47** [comment]
The `BootstrapExperiment.GenerateSelection` simplifies the scripting for #1056.
Suppose that `sampleX` and `sampleY` are defined, with same size `sample_size`.
It replaces the code:
```
# Create a sample with (X,Y)
fullSample = ot.Sample(sample_size, dimensionX + 1)
fullSample[:, 0:dimensionX] = sampleX
fullSample[:, dimensionX] = sampleY
# Create SRC bootstrap sample
bootstrap_size = 500
SRC_bootstrap = ot.Sample(bootstrap_size, dimensionX)
bootstrap = ot.BootstrapExperiment(fullSample)
for i in range(bootstrap_size):
    fullSampleBS = bootstrap.generate()
    sampleXBS = fullSampleBS[:, 0:dimensionX]
    sampleYBS = fullSampleBS[:, dimensionX]
    SRC_bootstrap[i, :] = ot.CorrelationAnalysis.SRC(sampleXBS, sampleYBS)
```
with:
```
bootstrap_size = 500
SRC_bootstrap = ot.Sample(bootstrap_size, dimensionX)
for i in range(bootstrap_size):
    indices = ot.BootstrapExperiment.GenerateSelection(sample_size, sample_size)
    sampleX_bootstrap, sampleY_bootstrap = sampleX[indices], sampleY[indices]
    SRC_bootstrap[i, :] = ot.CorrelationAnalysis.SRC(sampleX_bootstrap, sampleY_bootstrap)
```
This is minimal.

File: lib/src/Uncertainty/Algorithm/WeightedExperiments/LeaveOneOutSplitter.cxx [hunk 1]
```diff
@@ -0,0 +1,96 @@
+//                                               -*- C++ -*-
+/**
+ *  @brief LeaveOneOutSplitter
+ *
+ *  Copyright 2005-2020 Airbus-EDF-IMACS-ONERA-Phimeca
```
Comment (sofianehaddad) [inline lib/src/Uncertainty/Algorithm/WeightedExperiments/LeaveOneOutSplitter.cxx:5]: since you need to update to 2021 ;-)

**Review by sofianehaddad — COMMENTED** [review]
What is missing to make the PR tagged as ready?

**Comment by jschueller** [comment]
> What is missing to make the PR tagged as ready?

I think it should be okay, and maybe this pattern could benefit in other places

File: python/test/t_Splitter_std.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+splitters.append(splitter)
+splitter = ot.KFoldSplitter(N, 5)
+splitter.setRandomize(True)
+splitters.append(splitter)
+
+for splitter in splitters:
+    print("splitter=", splitter)
+    size = 0
+    for indices1, indices2 in splitter:
+        print('indices=', indices1, indices2)
```
Comment (josephmure) [inline python/test/t_Splitter_std.py:20]: Why this print? There is no corresponding `.expout` file.

**Review by josephmure — COMMENTED** [review]
There is no C++ test. Is this voluntary? Anyway, I think this will be a very handy class. Thanks @jschueller!

**Comment by josephmure** [comment]
> The `BootstrapExperiment.GenerateSelection` simplifies the scripting for #1056.
> Suppose that `sampleX` and `sampleY` are defined, with same size `sample_size`.
> It replaces the code:
> 
> *[...code...]*

@mbaudin47 Does snippet come from a doc example? If so, then this PR should probably change it in the way you suggest before it is merged.

File: python/src/KFoldSplitter_doc.i.in [hunk 3]
```diff
@@ -0,0 +1,33 @@
+%feature("docstring") OT::KFoldSplitter
+"K-fold splitter.
+
+Generates pairs of indices to split a set.
```
Comment (josephmure) [inline python/src/KFoldSplitter_doc.i.in:4]: Could you perhaps explain in further detail how the split is done?

File: python/src/LeaveOneOutSplitter_doc.i.in [hunk 4]
```diff
@@ -0,0 +1,22 @@
+%define OT_LeaveOneOutSplitter_doc
+"Leave-one-out splitter.
+
+Generates pairs of indices to split a set.
```
Comment (josephmure) [inline python/src/LeaveOneOutSplitter_doc.i.in:4]: Same.

File: python/test/t_Splitter_std.py (hunk 2 above)
Comment (josephmure) [reply python/test/t_Splitter_std.py:20]: Seriously, I don't understand the purpose of this print.

**Review by sofianehaddad — COMMENTED** [review]
Minor cosmetic
If possible, a C++ test could be added
Question : do we really need implementation/interface pattern? Objects seem "small"
Also I think we could not use interface as `iterator`

**Comment by jschueller** [comment]
> Minor cosmetic
> If possible, a C++ test could be added
> Question : do we really need implementation/interface pattern? Objects seem "small"
indeed, this is not necessary immediately, I can change that

> Also I think we could not use interface as `iterator`
what do you propose then ?

**Comment by sofianehaddad** [comment]
@jschueller 

> > Minor cosmetic
> > If possible, a C++ test could be added
> > Question : do we really need implementation/interface pattern? Objects seem "small"
> > indeed, this is not necessary immediately, I can change that
> 
> > Also I think we could not use interface as `iterator`
> > what do you propose then ?

`iterator` is ok. I mean the following code might cause troubles

```
spiltter = ot.Splitter(ot.KFold(...))
for indices1, indices2 in splitter:
    x_train, x_test = x[indices1], x[indices2] 
```
There is no overload of the `Splitter` interface class with `next/iter` methods

**Comment by jschueller** [comment]
oh, yes, I guess if I removed the interface it will go away

**Review by sofianehaddad — APPROVED** [review]

**Comment by jschueller** [comment]
todo: bernstein, resourcemap

**Review by regislebrun — APPROVED** [review]

**Pull request merged**